### PR TITLE
Refactor `merkle_tree::block`

### PR DESCRIPTION
**Pull request by tcharding**
PR 2 in the `merkle_tree::block` series, used to be on top of the now merged #1374.

Do a bunch of refactorings in preparation for more invasive changes. This is a separate PR because, other than the first patch which moves hex strings to `tests/data/` the other patches are refactoring only patches, no logic changes. However the last patch is big and will be annoying to review - sorry about that. If you really oppose this basically stylistic patch putting important things first, the opposite of C code, please say and I'll try to stop doing it.

### CODE REVIEW

**Review by apoelstra — DISMISSED** [review]
ACK f9f20285bc62906207f0b1a5f3ddf9724441f27d

**Comment by tcharding** [comment]
Rebased and redid the last patch to use the formatting changes picked up in rebase.

**Review by apoelstra — APPROVED** [review]
ACK 676ed10d0e4d50b99f030927538073dfd815755b

File: bitcoin/src/merkle_tree/block.rs [hunk 1]
```diff
@@ -42,52 +42,123 @@
...
-            NoTransactions => write!(f, "partial merkle tree contains no transactions"),
-            TooManyTransactions => write!(f, "too many transactions"),
-            BadFormat(ref s) => write!(f, "general format error: {}", s),
+    /// Create a MerkleBlock from the block's header and txids, that contain proofs for specific txids.
+    ///
+    /// The `header` is the block header, `block_txids` is the full list of txids included in the block and
+    /// `match_txids` is a function that returns true for the ids that should be included in the partial merkle tree.
+    pub fn from_header_txids_with_predicate<F>(
+        header: &block::Header,
+        block_txids: &[Txid],
```
Comment (hashmap) [inline bitcoin/src/merkle_tree/block.rs:119]: should we accept iterator here? we allocate a vec above just to use iterator API to allocate another vec below

File: bitcoin/src/merkle_tree/block.rs [hunk 2]
```diff
@@ -42,52 +42,123 @@
...
+        match_txids: F,
+    ) -> Self
+    where
+        F: Fn(&Txid) -> bool,
+    {
+        let matches: Vec<bool> = block_txids.iter().map(match_txids).collect();
+
+        let pmt = PartialMerkleTree::from_txids(block_txids, &matches);
```
Comment (hashmap) [inline bitcoin/src/merkle_tree/block.rs:127]: As understand it's out of the scope of this PR but do we plan to change `from_txids` to return `Result`? I think people prefer panic-free public APIs.
Another option would be to accept a slice/iterator of tuples `(TxId, bool)` - perhaps wrapped into a newtype, so an inconsistent state would be impossible to create

File: bitcoin/src/merkle_tree/block.rs (hunk 1 above)
Comment (Kixunil) [reply bitcoin/src/merkle_tree/block.rs:119]: Yeah, I was looking at it and the API is not great. But it's unrelated to this PR.

**Comment by tcharding** [comment]
Thanks for the review @hashmap! I created an issue to make sure your suggestions are followed up on. Thanks again.

**Review by apoelstra — APPROVED** [review]
ACK 55ddc3288ee0d877b83d9c69cd27cbdcffba187a

**Comment by tcharding** [comment]
Rebased. I had to redo the last patch manually to make sure there were no rebase mistakes so I spit it into two patches to make the diff easier to read - both are just moving code, no logic changes.

**Review by Kixunil — APPROVED** [review]
ACK 861fdd6ab13b8d13938d439cb90109e0e0d51341

**Review by apoelstra — APPROVED** [review]
ACK 861fdd6ab13b8d13938d439cb90109e0e0d51341

**Pull request merged**